camel-core/src/main/java/org/apache/camel/util/MessageHelper.java

```diff
@@ -546,7 +546,10 @@ public static String doDumpMessageHistoryStacktrace(Exchange exchange, ExchangeF
         sb.append("\n");
         sb.append("Message History\n");
         sb.append("---------------------------------------------------------------------------------------------------------------------------------------\n");
-        sb.append(String.format(MESSAGE_HISTORY_HEADER, "RouteId", "ProcessorId", "Processor", "Elapsed (ms)"));
+        String goMessageHistoryHeaeder = exchange.getContext().getGlobalOption(Exchange.MESSAGE_HISTORY_HEADER);
+        sb.append(String.format(
+                         goMessageHistoryHeaeder == null ? MESSAGE_HISTORY_HEADER : goMessageHistoryHeaeder,
+                         "RouteId", "ProcessorId", "Processor", "Elapsed (ms)"));
```

Comment:
I would prefer to write this as:

    final CamelContext context = exchange.getContext();
    final String historyHeaderFormat = Optional.ofNullable(context.getGlobalOption(Exchange.MESSAGE_HISTORY_HEADER).orElse(MESSAGE_HISTORY_HEADER);
    sb.append(String.format(historyHeaderFormat, "RouteId", "ProcessorId", "Processor", "Elapsed (ms)"));

This way it's a bit cleaner.